luarules/gadgets/unit_shield_behaviour.lua

```diff
@@ -0,0 +1,350 @@
+		local unitsTable = spGetUnitsInSphere(x, y, z, (shieldData.radius - 10))
+
+		for _, unitID in ipairs(unitsTable) do
+			if shieldedUnits[unitID] then
+				shieldedUnits[unitID][shieldUnitID] = true
+			else
+				shieldedUnits[unitID] = {}
+				shieldedUnits[unitID][shieldUnitID] = true
+			end
+		end
+
+		shieldData.shieldCoverageChecked = true
+	end
+end
+
+local function shieldNegatesDamageCheck(unitID, unitTeam, attackerID, attackerTeam)
+	if shieldedUnits[unitID] and allTeams[unitTeam] ~= allTeams[attackerTeam] and next(shieldedUnits[unitID]) then
+		if not shieldedUnits[attackerID] or (not next(shieldedUnits[attackerID]) and next(shieldedUnits[unitID])) then
+			return true
+		end
+
+		for subKey in pairs(shieldedUnits[unitID]) do
+			if shieldedUnits[attackerID][subKey] then
+				break
+			else
+				--The units have to share all of the same shield spaces. As soon as a mismatch is found, that means they don't occupy the same shield space and the shot should be blocked.
+				return true
+			end
+		end
+	end
+
+	return false
+end
+
+local shieldUnitsTotalCount = 0
+local shieldUnitIndex = {}
+local shieldCheckFlags = {}
+local lastShieldCheckChunkNumber = 1
+local shieldCheckChunkSize = 10
+local shieldCheckEndIndex = 1
+
+function gadget:GameFrame(frame)
+	gameSeconds = spGetGameSeconds()
+
+	for shieldUnitID in pairs(shieldCheckFlags) do
+		local shieldData = shieldUnitsData[shieldUnitID]
+
+		if shieldData then
+			if shieldData.shieldDamage > 0 and shieldData.shieldWeaponNumber > -1 then
+				local enabledState, shieldPower = spGetUnitShieldState(shieldUnitID)
+				shieldPower = shieldPower - shieldData.shieldDamage
+
+				if shieldPower < 0 then
+					shieldPower = 0
+				end
+
+				spSetUnitShieldState(shieldUnitID, shieldData.shieldWeaponNumber, shieldPower)
+				shieldData.shieldDamage = 0
+
+				if shieldData.downtimeReset < gameSeconds and shieldPower <= 0 then
+					triggerDowntime(shieldUnitID, shieldData.shieldWeaponNumber)
+					removeCoveredUnits(shieldUnitID)
+				end
+			else
+				shieldData.shieldDamage = 0
+			end
+
+			shieldCheckFlags[shieldUnitID] = nil
+		end
+
+	end
+	for shieldUnitID, shieldData in pairs(shieldUnitsData) do
+		if frame % Game.gameSpeed == 0 then
+			if select(2, spGetUnitIsStunned(shieldUnitID)) and shieldData.downtimeReset ~= 0 then
+				shieldData.downtimeReset = shieldData.downtimeReset + 1
+			end
+		end
+
+		if frame % 10 == 0 then
+
+			if shieldData.shieldEnabled == false and shieldData.downtimeReset ~= 0 and shieldData.downtimeReset <= gameSeconds then
+				shieldData.downtimeReset = 0
+				shieldData.shieldEnabled = true
+				spSetUnitShieldRechargeDelay(shieldUnitID, shieldData.shieldWeaponNumber, 0)
+
+				-- This section is to allow slower moving projectiles already inside the shield when it comes back online to damage units within the radius.
+				local x, y, z = spGetUnitPosition(shieldUnitID)
+				local radius = shieldData.radius-25
```

Comment:
What does the 25 represent? Why was this value chosen?

Reply:
Shouldn't this be the same value of 10 used for shield protection exclusion?

Reply:
The 25 was calibrated to prevent larger units partially sticking out of The shield taking direct hits but ignoring damage because they're technically inside the radius. This was done through trial and error in real world tests

The 10 was calibrated around what AOE is able to reach inside of the shield in a problematic way. If the radius is small enough then I don't need to worry about damage and I can save the performance impact. 

Perhaps they should be defined as a local variable instead of a magic number. I advise that you leave the actual values as they are.

Reply:
But this 25 is for projectiles that are already inside the shield when it comes back on not being intercepted. Why is it needed at all? Why are projectiles inside the shield near the edge different?

(Also it's [squaring the circle](https://en.wikipedia.org/wiki/Squaring_the_circle) because the engine only provides a way to get projectiles inside a square, but the shield is a circle, so there will be some funniness at the corners.)

Reply:
Oh sorry I was thinking of a different line of code. I'm trying to answer quickly while juggling other stuff.

This particular offset is to get the square roughly within the boundaries of the shield to reduce that corner funniness you were talking about in the parentheses. 

This exemption to damage mitigation is important for slower moving projectiles like the banisher missile getting inside of the shield while it is offline. When that shield turns back on it immediately protects the units underneath and would mitigate damage because the banisher is outside the shield radius even though the projectile made it inside of the shield already. 

The offset may or may not be removed. I thought it more important to have the box kind of fit inside of the shield radius instead of overlapping outside of it.

Reply:
In that case it should be calculated from the shield radius, a hardcoded value will only work with one particular radius.

Reply:
Do radius*.45 for equivalent adjustable value maybe
Or .5, not a significant difference